File: stone_lib/src/main/java/com/github/klee0kai/stone/closed/IModule.java
package com.github.klee0kai.stone.closed;

import com.github.klee0kai.stone.closed.types.SwitchCacheParam;

import java.util.Set;

/**
 * Stone Private class
 */
public interface IModule {

    /**
     * Init module
     *
     * @param ob can be:
     *           - a factory instance
     *           - a factory class
     */
    boolean init(Object ob);


    /**
     * Init caches from module prototype.
     * using in extOf method
     *
     * @param onlyBindInstance update cached only for bindInstance variables
     */
    void initCachesFrom(IModule module, boolean onlyBindInstance);

    /**
     * bind instance objects
     *
     * @param object - An instance of bindable objects
     */
    boolean bind(Object object);

    /**
     * get component's factory
     *
     * @return
     */
    Object getFactory();

    /**
     * Switch cache type for scope
     */
    void switchRef(Set<Class> scopes, SwitchCacheParam param);

}

File: stone_multiplatform/src/commonMain/kotlin/com/github/klee0kai/stone/__hidden__/IModule.kt
package com.github.klee0kai.stone.__hidden__

import kotlin.reflect.KClass

/**
 * Stone Private class
 */
interface IModule {

    /**
     * get component's factory
     *
     * @return
     */
    val factory: Any?

    /**
     * Init module
     *
     * @param ob can be:
     * - a factory instance
     * - a factory class
     */
    fun __init(ob: Any): Boolean

    /**
     * Init caches from module prototype.
     * using in extOf method
     */
    fun __initCachesFrom(module: IModule)

    /**
     * Update values of bindInstance variables
     *
     * @param module related module, source to update
     */
    fun __updateBindInstancesFrom(module: IModule)

    /**
     * bind instance objects
     *
     * @param object - An instance of bindable objects
     */
    fun __bind(`object`: Any?): Boolean


    /**
     * Switch cache type for scope
     */
    public fun __switchRef(scopes: Set<KClass<*>>, __params: SwitchCacheParam)

    /**
     * Clear null refs.
     * Useful after gc
     */
    fun __clearNulls()
}
